plugin_test.go

```diff
@@ -0,0 +1,196 @@
+package main
+
+import (
+	"bytes"
+	"context"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"path"
+	"strings"
+	"testing"
+	"time"
+
+	"google.golang.org/api/option"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+	"google.golang.org/grpc/test/bufconn"
+
+	secretmanager "cloud.google.com/go/secretmanager/apiv1"
+	secretmanagerpb "google.golang.org/genproto/googleapis/cloud/secretmanager/v1"
+)
+
+var sampleAttrs = `
+{
+	"secrets": "array:\n  - |\n    resourceName: \"projects/project/secrets/test/versions/latest\"\n    fileName: \"good1.txt\"\n"
+}
+`
+
+func TestPlugin(t *testing.T) {
+	dir, cleanup := driveMountHelper(t)
+	defer cleanup()
+
+	want := []byte("My Secret")
+	client, shutdown := mock(t, &mockSecretServer{
+		accessFn: func(ctx context.Context, _ *secretmanagerpb.AccessSecretVersionRequest) (*secretmanagerpb.AccessSecretVersionResponse, error) {
+			return &secretmanagerpb.AccessSecretVersionResponse{
+				Payload: &secretmanagerpb.SecretPayload{
+					Data: want,
+				},
+			}, nil
+		},
+	})
+	defer shutdown()
+
+	if got := plugin(context.Background(), client, sampleAttrs, "{}", dir, "777"); got != nil {
+		t.Errorf("plugin() got err = %v, want err = nil", got)
+	}
+
+	got, err := ioutil.ReadFile(path.Join(dir, "good1.txt"))
+	if err != nil {
+		t.Errorf("error reading secret. got err = %v, want err = nil", err)
+	}
+	if !bytes.Equal(got, want) {
+		t.Errorf("plugin() wrote unexpected secret value. got = %v, want = %v", got, want)
+	}
+}
+
+func TestPluginSMError(t *testing.T) {
+	dir, cleanup := driveMountHelper(t)
+	defer cleanup()
+
+	client, shutdown := mock(t, &mockSecretServer{
+		accessFn: func(ctx context.Context, _ *secretmanagerpb.AccessSecretVersionRequest) (*secretmanagerpb.AccessSecretVersionResponse, error) {
+			return nil, status.Error(codes.FailedPrecondition, "Secret is Disabled")
+		},
+	})
+	defer shutdown()
+
+	got := plugin(context.Background(), client, sampleAttrs, "{}", dir, "777")
+	if !strings.Contains(got.Error(), "FailedPrecondition") {
+		t.Errorf("plugin() got err = %v, want err = nil", got)
+
+	}
+}
+
+func TestPluginConfigErrors(t *testing.T) {
+	dir, cleanup := driveMountHelper(t)
+	defer cleanup()
+
+	client, shutdown := mock(t, &mockSecretServer{
+		accessFn: func(ctx context.Context, _ *secretmanagerpb.AccessSecretVersionRequest) (*secretmanagerpb.AccessSecretVersionResponse, error) {
+			return &secretmanagerpb.AccessSecretVersionResponse{
+				Payload: &secretmanagerpb.SecretPayload{
+					Data: []byte("data"),
+				},
+			}, nil
+		},
+	})
+	defer shutdown()
+
+	tests := []struct {
+		name    string
+		attrs   string
+		secrets string
+		dir     string
+		perm    string
+	}{
+		{
+			name:    "unparsable attributes",
+			attrs:   "",
+			secrets: "{}",
+			dir:     dir,
+			perm:    "777",
```

Comment:
Can you add a test for umask too?

Reply:
switched to an int flag with a default value, so no longer part of the unit test